DEVELOPMENT.md

```diff
@@ -87,6 +87,15 @@ Maintain the same quality for production and test code.
 
 When writing a Git commit message, follow these [guidelines](https://chris.beams.io/posts/git-commit/).
 
+### Git merge strategy
+
+We prefer to `rebase and merge`, so cleaning up and squashing commits on the feature branch is best.
+We care a lot about the history, so this is why how change is structured between commits and what the
+commit messages are is also a part of code review. However we prefer not to merge anything where changes go
+back and forth between commits in the same PR. This makes the entire git history much cleaner. Notice that
+writing code is a creative process and it is natural to go back and forth when writing code, however
+when visiting history such small steps are not helpful.
```

Comment:
Here's my stab at a rewording:

```
### Git merge strategy
PRs are usually merged into `master` using the  `rebase and merge` strategy. 

It is important to keep commits on feature branches neat, squashing the feature branch as necessary.
Commit messages and  history are also important, as they are used by other developers to keep track of the motivation behind changes.
Keep logical diffs grouped together in separate commits, and order commits in a way that explains
the progress of the changes. Rewriting and reordering commits may be a necessary part of the PR review
process, as the code changes.
```

Reply:
This better .. I would go even further and mention that typical PRs should strive to just contain one commit.. 

I am not aware of an actual policy beyond that .. since "neat" and "logical" and such are all very subjective terms and are not that actionable.

Reply:
> typical PRs should strive to just contain one commit..


that can cause people to create too big commits, let's not write that

Reply:
> typical PRs should strive to just contain one commit..

I don't think that's the case. Contrarily, we often ask contributors to extract unrelated changes (e.g. fix-typos, reformats) to separate commits.
The rule is that a commit should be a logically separate chunk of work. 

One thing that could be probably mentioned is that if a PR consists of multiple commits, it is expected that every prefix of it is  _correct_. That is, there might be preparatory commits at the bottom of the stack that don't bring any value by themselves, but none of the commits should introduce an error that is fixed by some future commit.

Reply:
> it is expected that every prefix of it is correct

I like that rule 👍